efficientdet/aug/mosaic.py

```diff
@@ -0,0 +1,261 @@
+        """
+        # TODO #MED #use n_images to build mosaic.
+        self._n_images = n_images
+        self._out_size = out_size
+        self._minimum_mosaic_image_dim = _minimum_mosaic_image_dim
+        assert (
+            _minimum_mosaic_image_dim > 0
+        ), "Minimum Mosaic image dimension should be above 0"
+
+    @property
+    def n_images(self) -> int:
+        """n_images.
+            number of images tot mosaic property.
+        """
+        return self._n_images
+
+    @property
+    def out_size(self) -> int:
+        """out_size.
+            Output image size property.
+        """
+        return self._out_size
+
+    def _mosaic_divide_points(self) -> Tuple:
+        """_mosaic_divide_points.
+            Returns:
+                tuple of x and y which corresponds to mosaic divide points.
+        """
+        x_point = tf.random.uniform(
+            shape=[1],
+            minval=tf.cast(
+                self.out_size[0] * (self._minimum_mosaic_image_dim / 100), tf.int32
+            ),
+            maxval=tf.cast(
+                self.out_size[0] * ((100 - self._minimum_mosaic_image_dim) / 100),
+                tf.int32,
+            ),
+            dtype=tf.int32,
+        )
+        y_point = tf.random.uniform(
+            shape=[1],
+            minval=tf.cast(
+                self.out_size[1] * (self._minimum_mosaic_image_dim / 100), tf.int32
+            ),
+            maxval=tf.cast(
+                self.out_size[1] * ((100 - self._minimum_mosaic_image_dim) / 100),
+                tf.int32,
+            ),
+            dtype=tf.int32,
+        )
+        return x_point, y_point
+
+    @staticmethod
+    def _scale_box(box, image, mosaic_image):
+        """_scale_box.
+            static bounding boxes scaling methods which scales the boxes with mosaic sub image.
+
+        Args:
+            box: mosaic image box.
+            image: original image.
+            mosaic_image: mosaic sub image.
+        Returns:
+            Scaled bounding boxes.
+
+        """
+        return [
+            box[0] * tf.shape(mosaic_image)[1] / tf.shape(image)[1],
+            box[1] * tf.shape(mosaic_image)[0] / tf.shape(image)[0],
+            box[2] * tf.shape(mosaic_image)[1] / tf.shape(image)[1],
+            box[-1] * tf.shape(mosaic_image)[0] / tf.shape(image)[0],
+        ]
+
+    def _scale_images(self, images, mosaic_divide_points: Tuple) -> Tuple:
+        """_mosaic.
+            Scale Sub Images.
+
+        Args:
+            images: original single images to make mosaic.
+            mosaic_divide_points: Points to build mosaic around on given output size.
+        Returns:
+            (tuple)
+            Scaled Mosaic sub images.
+        """
+        x, y = mosaic_divide_points[0][0], mosaic_divide_points[1][0]
```

Comment:
The mosaic algorithm doesn't do this resizing - it does cropping.  See the darknet implementation [here](https://github.com/AlexeyAB/darknet/blob/master/src/data.c#L1907).

Reply:
@LucasSloan we can make mosaic of original image and then crop the mosaic with output_size

Reply:
@LucasSloan FYI ,this is initial version of mosaic we can test with resizing ,if it doesnt help ,I can surely make the cropping version.

Reply:
Resizing is already a knob that we have in the augmentation toolkit here.  If you do mosiac via resizing, you're reducing the effective size of everything by a factor of 2.  It's not good to have this interaction between different augmentations - please redo this as a crop.

Reply:
you can control the mosaic image shape via output_size

Reply:
so lets say ,u have 4 images with shape (512x512) you can make a mosaic of them with output size (512,512) ,right?